feat: expose agent stats via Prometheus endpoint

mtojek opened the pull request:
Related: https://github.com/coder/coder/issues/6724

This PR exposes agent stats stored in the database via the Prometheus endpoint.

```
# HELP coderd_agentstats_connection_count The number of established connections by agent
# TYPE coderd_agentstats_connection_count gauge
coderd_agentstats_connection_count{agent_name="main",username="admin",workspace_name="workspace1"} 2
# HELP coderd_agentstats_connection_median_latency The median agent connection latency
# TYPE coderd_agentstats_connection_median_latency gauge
coderd_agentstats_connection_median_latency{agent_name="main",username="admin",workspace_name="workspace1"} 0.002476
# HELP coderd_agentstats_rx_bytes Agent Rx bytes
# TYPE coderd_agentstats_rx_bytes gauge
coderd_agentstats_rx_bytes{agent_name="main",username="admin",workspace_name="workspace1"} 7726
# HELP coderd_agentstats_session_count_jetbrains The number of session established by JetBrains
# TYPE coderd_agentstats_session_count_jetbrains gauge
coderd_agentstats_session_count_jetbrains{agent_name="main",username="admin",workspace_name="workspace1"} 0
# HELP coderd_agentstats_session_count_reconnecting_pty The number of session established by reconnecting PTY
# TYPE coderd_agentstats_session_count_reconnecting_pty gauge
coderd_agentstats_session_count_reconnecting_pty{agent_name="main",username="admin",workspace_name="workspace1"} 1
# HELP coderd_agentstats_session_count_ssh The number of session established by SSH
# TYPE coderd_agentstats_session_count_ssh gauge
coderd_agentstats_session_count_ssh{agent_name="main",username="admin",workspace_name="workspace1"} 0
# HELP coderd_agentstats_session_count_vscode The number of session established by VSCode
# TYPE coderd_agentstats_session_count_vscode gauge
coderd_agentstats_session_count_vscode{agent_name="main",username="admin",workspace_name="workspace1"} 0
# HELP coderd_agentstats_tx_bytes Agent Tx bytes
# TYPE coderd_agentstats_tx_bytes gauge
coderd_agentstats_tx_bytes{agent_name="main",username="admin",workspace_name="workspace1"} 6623
```

mtojek committed:
WIP

mtojek committed:
Merge branch 'main' into 6724-metrics

mtojek committed:
WIP

mtojek committed:
WIP

mtojek committed:
Agents

mtojek committed:
fix

mtojek committed:
1min

mtojek committed:
fix

mtojek committed:
WIP

mtojek committed:
Test

mtojek committed:
docs

mtojek committed:
fmt

mtojek committed:
Add timer to measure the metrics collection

mtojek committed:
Use CachedGaugeVec

mtojek committed:
Unit tests

mtojek committed:
WIP

mtojek committed:
Merge branch 'main' into 6724-metrics-2

mtojek committed:
WIP

mtojek committed:
db: GetWorkspaceAgentStatsAndLabels

mtojek committed:
fmt

mtojek committed:
WIP

mtojek committed:
Merge branch 'main' into 6724-metrics-2

mtojek committed:
gauges

mtojek committed:
feat: collect

mtojek committed:
fix

mtojek committed:
fmt

deansheather commented on cli/server.go:
```diff
@@ -844,6 +844,12 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
 				}
 				defer closeWorkspacesFunc()
 
+				closeAgentStatsFunc, err := prometheusmetrics.AgentStats(ctx, logger, options.PrometheusRegistry, options.Database, time.Now(), 0)
```
I think this should be disableable via a flag. We add lots of stats here for every single agent, and cloud metrics providers often charge per metric/datapoint so the costs can rack up quickly.

mtojek replied:
Flag added

deansheather commented on coderd/database/queries/workspaceagentstats.sql:
```diff
@@ -103,3 +103,38 @@ WITH agent_stats AS (
+	WHERE a.rn = 1
+	GROUP BY a.user_id, a.agent_id, a.workspace_id
+)
+SELECT users.username, workspace_agents.name AS agent_name, workspaces.name AS workspace_name, workspace_rx_bytes, workspace_tx_bytes, session_count_vscode, session_count_ssh, session_count_jetbrains, session_count_reconnecting_pty, connection_count, connection_median_latency_ms
+FROM agent_stats JOIN latest_agent_stats ON agent_stats.agent_id = latest_agent_stats.agent_id
+JOIN users ON users.id = agent_stats.user_id
+JOIN workspace_agents ON workspace_agents.id = agent_stats.agent_id
+JOIN workspaces ON workspaces.id = agent_stats.workspace_id;
```
This last part of the query could be formatted in an easier to read way like you have above.

mtojek replied:
Formatted

deansheather commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -283,9 +283,175 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
+			logger.Debug(ctx, "Agent metrics collection is done")
+			metricsCollectorAgentStats.Observe(timer.ObserveDuration().Seconds())
+
+			createdAfter = checkpoint
```
Should call ticker.Reset() so it's `duration` between stats attempts instead of a fixed time between attempts starting. Other places in the code do this too like apphealth, dbpurge, etc.

mtojek replied:
Good call, done

deansheather commented on coderd/prometheusmetrics/prometheusmetrics_test.go:
```diff
@@ -352,3 +353,85 @@ func TestAgents(t *testing.T) {
+	t.Cleanup(cancel)
+
+	// when
+	_, err = agentClient.PostStats(context.Background(), &agentsdk.Stats{
```
I think the test should have two+ agents and multiple posted metrics for each to see how the queries can handle it.

mtojek replied:
Done. Now we have 3 agents and some random numbers. There is also a golden file to match the results.

mtojek committed:
minor fixes

mtojek committed:
Prometheus flag

mtojek committed:
fix

mafredri commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -200,7 +200,7 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
 			if err != nil {
 				logger.Error(ctx, "can't get workspace rows", slog.Error(err))
-				continue
+				goto done
```
Not a blocker, but prefer to avoid goto when possible. Introduces a more complex flow.

In this instance e.g. an if/else might be preferable even though it introduces indentation.

mtojek replied:
Fixed, although I prefer "goto" in this situation :)

mafredri commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -283,8 +283,177 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
+			ticker.Reset(duration)
+		}
+	}()
+	return cancelFunc, nil
```
I believe this function is considered "close" elsewhere, but doesn't give a guarantee that the service has closed. This could e.g. leave running go routines in tests which would be caught by goleak.

(Same with the function below)

It might never be an actual problem but more of a semantic/foolproofing thing. It's never a guarantee though that a function taking a context will exit immediate on cancellation.

mtojek replied:
Would you like me to rename the calling place: `closeAgentStatsFunc` -> `cancelAgentStatsFunc`?

BTW it affects also other functions: `closeUsersFunc`, `closeWorkspacesFunc`, `closeAgentsFunc`

mtojek replied:
Ok, I applied the pattern to all metrics-related functions.

mafredri commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -283,8 +283,177 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
+				goto done
+			}
+
+			if len(stats) == 0 {
```
This check could guard the commits below instead, which would avoid the need for goto.

mtojek replied:
Fixed

mafredri reviewed (approved):
Looks good, had a few comments and observations but nothing critical.

mtojek commented on cli/testdata/coder_server_--help.golden:
```diff
@@ -90,6 +90,9 @@ Use a YAML configuration file when your server launch become unwieldy.
       --prometheus-address host:port, $CODER_PROMETHEUS_ADDRESS (default: 127.0.0.1:2112)
           The bind address to serve prometheus metrics.
 
+      --prometheus-collect-agent-stats bool, $CODER_PROMETHEUS_COLLECT_AGENT_STATS
```
I would say the opposite. It should be disabled by default so that when admin enables Prometheus metrics, they won't be extra charged. BTW I don't have a strong preference here, I'm ok with adjusting this behavior.

mafredri replied:
I’m alright with leaving it disabled by default. I don’t really have a concept for how much extra data/load this introduces so if it makes sense to keep disabled by default let’s not change it.

mtojek commented on coderd/database/queries/workspaceagentstats.sql:
```diff
@@ -103,3 +103,55 @@ WITH agent_stats AS (
+		user_id,
+		agent_id,
+		workspace_id,
+		coalesce(SUM(rx_bytes), 0)::bigint AS workspace_rx_bytes,
```
It is per-agent and per-workspace, so per-agent is narrower. I will rename the column.

mtojek replied:
Fixed

mtojek committed:
WIP

mtojek committed:
fix tests

mtojek committed:
WIP

mtojek committed:
fix json

mtojek committed:
Rx Tx bytes

mtojek commented on coderd/database/dbfake/databasefake.go:
```diff
@@ -3998,6 +3998,77 @@ func (q *fakeQuerier) GetWorkspaceAgentStats(_ context.Context, createdAfter tim
+
+		stat.Username = user.Username
+
+		workspace, err := q.GetWorkspaceByID(ctx, agentStat.WorkspaceID)
```
It is RWMutex, and we're using here only RLocks, so I guess that is's safe?

mtojek committed:
CloseFunc

mtojek committed:
fix

mtojek committed:
fix

mtojek committed:
Fixes

mtojek committed:
fix

mtojek commented:
@mafredri I addressed your suggestions regarding cancellation vs. closing, but now the code is flaky. I'm not sure what is wrong here: https://github.com/coder/coder/actions/runs/4699908242/jobs/8333944515

Would you mind throwing an eye on the output? It looks like it complains about dbauthz.

mtojek committed:
fix: IgnoreErrors

mtojek committed:
Fix: Windows

mtojek committed:
fix

mtojek committed:
reflect.DeepEquals

mtojek commented:
Ok, it looks like it's fixed now 👍
